### DISCUSSION TITLE
Allow pages to be "API Routes"

### DISCUSSION THREAD

**Discussion by egoist** [vercel/next.js#25710] [Ideas]
Currently you can use `getServerSideProps` to eliminate the need for an additional API for fetching data, but API is still needed for "updating data", my proposal is to allow handlers for `POST`, `PUT`, and other requests in pages:

```tsx
// pages/new-todo.tsx
import { useForm, rules } from 'next'

export const handler = async ({ req }) => {
  const data = await req.validate({
    title: rules.string().required(),
  })
  const todo = await orm.todos.create({ title: data.title })
  return {
    redirect: `/todos/${todo.id}`,
  }
}

export default function Page() {
  const form = useForm({ values: { title: '' } })
  return (
    <form onSubmit={form.handleSubmit}>
      <input
        name="title"
        value={form.values.title}
        onChange={form.handleChange}
      />
      {form.errors.title && <div>{form.errors.title}</div>}
      <button type="submit">Create todo</button>
    </form>
  )
}
```

Inspired by [Inertia](https://inertiajs.com/forms) and [Remix](https://docs.remix.run/v0.17/tutorial/6-actions/).

**Comment by kvnwolf** [comment]
+1 on this. Would be really helpful to avoid creating another file at another location for a one-off DB write operation.

**Comment by anneau** [comment]
I use [next-runtime](https://github.com/smeijer/next-runtime) to resolve this.

But, I want you to support this officially
